run-command.c

```diff
@@ -966,18 +966,24 @@ int start_command(struct child_process *cmd)
 		return -1;
 	}
 
-	if (need_in)
+	if (need_in) {
 		close(fdin[0]);
+		set_cloexec(fdin[1]);
+	}
```

Comment:
Makes tons of sense to me.

Please note that there is a slight window here between the spawn and the `set_cloexec()`. I *think* we could reduce that window by inserting those `set_cloexec()` calls already in lines 693, 709 and 731, respectively.

This would *still* leave a slight window, especially when we keep in mind that the read-object hook might be spawned in some multi-threaded part of Git's code. But it would be a much, much smaller window.

If you do that, however, you will have to change the `close_pair()` calls after `fork()` returned  `0`, i.e. lines 799, 809 and 821.

Of course, the `set_cloexec()` call [amounts to a noop on Windows](https://github.com/git-for-windows/git/blob/v2.21.0.windows.1/compat/mingw.h#L142-L148). But then, it [already restricts which handles are inherited to stdin/stdout/stderr](https://github.com/git-for-windows/git/blob/v2.21.0.windows.1/compat/mingw.c#L1795-L1808).

Reply:
So I looked at this very carefully, and I think if **another** happens before `close(fdin[0]);` the spawned process will hang too. Thus I don't think moving this around will improve it by much.

Also - this flag must be set **after** the first fork but before the second one, so I think performing it in lines 693, 709 and 731 would not work.

Reply:
Thank you for the clarification!